crates/sui-indexer-alt-graphql/src/extensions/query_limits/mod.rs

```diff
@@ -0,0 +1,1149 @@
+    }
+}
+
+impl QueryLimitsChecker {
+    pub(crate) fn new(limits: QueryLimitsConfig, metrics: Arc<RpcMetrics>) -> Self {
+        Self {
+            limits: Arc::new(limits),
+            metrics,
+        }
+    }
+}
+
+impl ExtensionFactory for QueryLimitsChecker {
+    fn create(&self) -> Arc<dyn Extension> {
+        Arc::new(QueryLimitsCheckerExt {
+            limits: self.limits.clone(),
+            metrics: self.metrics.clone(),
+            doc: Mutex::new(None),
+            usage: Mutex::new(None),
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl Extension for QueryLimitsCheckerExt {
+    /// Performs initial checks about content length, and then stashes the parsed document so that
+    /// we can run our validation checks on it.
+    async fn parse_query(
+        &self,
+        ctx: &ExtensionContext<'_>,
+        query: &str,
+        variables: &Variables,
+        next: NextParseQuery<'_>,
+    ) -> ServerResult<ExecutableDocument> {
+        let &ContentLength(length) = ctx.data_unchecked();
+        if length > self.limits.max_payload_size() as u64 {
+            Err(Error::new_global(ErrorKind::PayloadSizeOverall {
+                limit: self.limits.max_payload_size(),
+                actual: length,
+            }))?;
+        }
+
+        let doc = next.run(ctx, query, variables).await?;
+
+        // Stash the parsed document so that we can run our validations on it, after the
+        // framework's validations have run.
+        *self.doc.lock().unwrap() = Some(ParsedDocument {
+            var: variables.clone(),
+            doc: doc.clone(),
+        });
+
+        Ok(doc)
+    }
+
+    /// Run the framework's validation rules, and then run our own validation rules to validate
+    /// that the query is within configurable limits before it is run, to protect the rest of the
+    /// system from doing too much work. Tests ensure that:
+    ///
+    /// - The query is not too large or too deep, as an AST.
+    /// - If the query is large, that does not translate into a lot of query work (it's okay to
+    ///   have large binary payloads to handle execution, but we don't want a query with a big
+    ///   footprint to translate into a query that requires a lot of work to execute).
+    /// - The query will not produce too large a response (estimated based on the upperbound number
+    ///   of output nodes that input query could produce).
+    async fn validation(
+        &self,
+        ctx: &ExtensionContext<'_>,
+        next: NextValidation<'_>,
+    ) -> Result<ValidationResult, Vec<ServerError>> {
+        let res = next.run(ctx).await?;
+
+        let Some(ParsedDocument { doc, var }) = self.doc.lock().unwrap().take() else {
+            return Ok(res);
+        };
+
+        let &ContentLength(length) = ctx.data_unchecked();
+        let pagination_config: &PaginationConfig = ctx.data_unchecked();
+
+        let _guard = self.metrics.limits_validation_latency.start_timer();
```

Comment:
I'm glad we have this metric. One concern I have is regarding the complexity of these check functions and whether they open up possibilities for dos attacks in case there's a bug that causes infinite loops somewhere. And since query limits checker is outside of timeout, does that mean a potential bug like that could cause the server to stall?

Reply:
Query limits is not outside of timeouts (see my earlier clarification)